daemon: Support the wildcard option for directRoutingDevice

ysksuzuki opened the pull request:
This PR implements support for the wildcard option for directRoutingDevice.

Fixes: #17852

Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Thanks for contributing!

ysksuzuki committed:
daemon: Support the wildcard option for directRoutingDevice

This implementation expands devices based on the concrete devices on the node,
sorts in increasing order and picks a first one,
when the wildcard option is specified.

Fixes: #17852

Signed-off-by: Yusuke Suzuki <yusuke-suzuki@cybozu.co.jp>

joamaki commented on daemon/cmd/devices.go:
```diff
@@ -283,37 +287,59 @@ func (dm *DeviceManager) updateDevicesFromRoutes(l3DevOK bool, routes []netlink.
-		return fmt.Errorf("Device wildcard expansion failed to detect devices. Please verify --%s option.",
-			option.Devices)
+		return nil, fmt.Errorf("Device wildcard expansion failed to detect devices. Please verify --%s option.",
+			devices)
```
`option.Devices` is the name of the argument, e.g. `option.Devices` is `devices`. Now this will print `Please verify --[eth+] option`, which is not what we want. You should pass in `option.Devices` or `option.DirectRoutingDevice` as an argument to this function so it'll show what argument was specified wrong.

joamaki commented on daemon/cmd/devices.go:
```diff
@@ -283,37 +287,59 @@ func (dm *DeviceManager) updateDevicesFromRoutes(l3DevOK bool, routes []netlink.
+	return nil
+}
+
+func expand(devices []string) ([]string, error) {
```
perhaps we could call this "expandDeviceWildcards" so it's more descriptive?

joamaki reviewed (changes requested):
Thanks! It's almost there, just couple small things...

joestringer reviewed (approved):
Docs changes LGTM. Looks like there's just some minor feedback from Jussi that should be addressed, then we can trigger the full testsuite.

ysksuzuki committed:
daemon: Fix expandDeviceWildcards function name and print an option name

Modify the function name appropriately, and print an option name
if something wrong with the wildcard expansion.

Signed-off-by: Yusuke Suzuki <yusuke-suzuki@cybozu.co.jp>

brb commented on daemon/cmd/devices.go:
```diff
@@ -283,37 +287,59 @@ func (dm *DeviceManager) updateDevicesFromRoutes(l3DevOK bool, routes []netlink.
+	if err != nil {
+		return err
+	}
+	option.Config.DirectRoutingDevice = expandedDevices[0]
```
We should panic if there is more than one device which matches the wildcard direct routing device match. Otherwise, it might lead to situations in which other than expected device has been selected.

ysksuzuki replied:
Thank you for your comment!

It doesn't work at least in our use case if cilium-agent panic if there is more than one device that matches the wildcard direct routing device match.

We have two types of nodes and one has devices eno1 and eno2, another has devices eno8303 and eno8403. Kubernetes internal IP is attached to a dummy interface on each node. In this case, we specify directRoutingDevice=eno+ and expect cilium-agent to pick eno1 or eno8303 accordingly.

I understand your concern. How about printing a warning log and showing users which device has been selected as the direct routing device if there is more than one device that matches the wildcard?

ysksuzuki replied:
@brb What do you think about this?

brb replied:
We are currently busy with the final bits of the upcoming v1.11 release. Let us back to you next week.

joamaki reviewed (approved):
Looks good to me once @brb's comment has been addressed.

brb reviewed (changes requested):
Could you mention in the docs about what direct routing dev is chosen if a wildcard matches many?

brb commented on Documentation/gettingstarted/kubeproxy-free.rst:
```diff
@@ -838,9 +838,12 @@ via ``devices`` then Cilium will use that device for direct routing.
+option, e.g. ``nodePort.directRoutingDevice=eth1``. The wildcard option can be
+used as well as the devices option, e.g. ``directRoutingDevice=eth+``.
+If more than one devices match the wildcard option, Cilium will sort them
+in increasing order and pick the first one. If the direct routing device does not
```
Nit: "increasing alphanumerical order"
